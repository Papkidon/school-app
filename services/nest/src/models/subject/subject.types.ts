import { z } from 'zod';

export const subjectCreateSchema = z.object({
  data: z.object({
    name: z.string(),
    points: z.number(),
  }),
});

export const subjectUpdateSchema = subjectCreateSchema.deepPartial();

export type subjectsWithClassesDto = {
  result: {
    id: string;
    name: string;
    classes: {
      id: string;
      number: string;
      max_population: number;
    }[];
  };
}[];

export type mappedSubjectsWithClassesDto = {
  id: string;
  name: string;
  classes: { id: string; number: string; max_population: number }[];
}[];

export type subjectType = {
  data: {
    name: string;
    points: number;
  };
};
