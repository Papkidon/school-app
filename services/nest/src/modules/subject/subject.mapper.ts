import {
  mappedSubjectsWithClassesDto,
  subjectsWithClassesDto,
} from '../../models/subject/subject.types';

export const mapSubjectsWithClasses = (
  subjects: subjectsWithClassesDto,
): mappedSubjectsWithClassesDto => {
  if (!subjects.length) return [];
  return subjects.map((subject) => {
    return {
      ...subject.result,
    };
  });
};
