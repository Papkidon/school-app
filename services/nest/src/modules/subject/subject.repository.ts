import { Injectable } from '@nestjs/common';
import { Subject } from '@prisma/client';
import { PrismaService } from 'nestjs-prisma';
import {
  subjectsWithClassesDto,
  subjectType,
} from '../../models/subject/subject.types';
import { DeepPartial } from '../../models/utility.types';

@Injectable()
export class SubjectRepository {
  constructor(private prisma: PrismaService) {}
  async findAll(): Promise<Subject[]> {
    return await this.prisma.subject.findMany();
  }

  async findById(id: string): Promise<Subject> {
    return await this.prisma.subject.findUnique({ where: { id } });
  }

  async findByName(name: string): Promise<Subject> {
    return await this.prisma.subject.findUnique({ where: { name } });
  }

  async findAllSubjectsWithClasses() {
    return (await this.prisma.$queryRaw`
        SELECT
        json_build_object (
            'id', s.id,
            'name', s.name,
            'points', s.points,
            'classes', jsonb_agg (c)
            ) AS result
       FROM "Subject" s
       LEFT JOIN "ClassSubject" as cs ON
       s.id = cs.subject_id
       LEFT JOIN "Class" as c ON
       cs.class_id = c.id
       GROUP BY s.id;`) as subjectsWithClassesDto;
  }

  async create(body: any): Promise<Subject> {
    return await this.prisma.subject.create(body);
  }

  async delete(id: string): Promise<Subject> {
    return await this.prisma.subject.delete({ where: { id } });
  }

  async update(id: string, body: DeepPartial<subjectType>): Promise<Subject> {
    return await this.prisma.subject.update({
      where: { id },
      data: { ...body.data },
    });
  }
}
