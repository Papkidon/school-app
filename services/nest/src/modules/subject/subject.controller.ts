import {
  Body,
  Controller,
  Delete,
  Get,
  Param,
  Post,
  Patch,
  HttpException,
  HttpStatus,
} from '@nestjs/common';
// eslint-disable-next-line @typescript-eslint/no-unused-vars
import { SubjectService } from './subject.service';
import { ZodValidationPipe } from '../../pipes/zod.validation.pipe';
import {
  subjectCreateSchema,
  subjectType,
  subjectUpdateSchema,
} from '../../models/subject/subject.types';
import { DeepPartial } from '../../models/utility.types';
import { Subject } from '@prisma/client';
import { mapSubjectsWithClasses } from './subject.mapper';
import { idSchema } from '../../models/utility.types';

@Controller('api/v1/subject')
export class SubjectController {
  constructor(private readonly subjectService: SubjectService) {}

  @Get('')
  async findAll(): Promise<Subject[]> {
    try {
      const result = await this.subjectService.findAll();

      if (!Object.keys(result).length) {
        throw new HttpException(
          'No user data exists in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return result;
    } catch (error) {
      throw error;
    }
  }

  @Get('classes')
  async findAllSubjectsWithClasses(): Promise<unknown> {
    try {
      return mapSubjectsWithClasses(
        await this.subjectService.findAllSubjectWithClasses(),
      );
    } catch (error) {
      throw error;
    }
  }

  @Post('')
  async create(
    @Body(new ZodValidationPipe(subjectCreateSchema))
    createSubjectDto: subjectType,
  ): Promise<Subject> {
    try {
      return await this.subjectService.create(createSubjectDto);
    } catch (error) {
      throw error;
    }
  }

  @Get(':id')
  async findById(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
  ): Promise<Subject> {
    try {
      const result = await this.subjectService.findById(id);

      if (!result) {
        throw new HttpException(
          'User with given id does not exist in the database',
          HttpStatus.NOT_FOUND,
        );
      }

      return result;
    } catch (error) {
      throw error;
    }
  }

  @Patch(':id')
  async update(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
    @Body(new ZodValidationPipe(subjectUpdateSchema))
    body: DeepPartial<subjectType>,
  ): Promise<Subject> {
    console.log(id);
    try {
      return await this.subjectService.update(id, body);
    } catch (error) {
      throw error;
    }
  }

  @Delete(':id')
  async delete(
    @Param('id', new ZodValidationPipe(idSchema)) id: string,
  ): Promise<Subject> {
    try {
      return await this.subjectService.delete(id);
    } catch (error) {
      throw error;
    }
  }
}
