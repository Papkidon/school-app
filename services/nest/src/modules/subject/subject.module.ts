import { Module } from '@nestjs/common';
import { PrismaService } from 'nestjs-prisma';
import { SubjectController } from './subject.controller';
import { SubjectRepository } from './subject.repository';
import { SubjectService } from './subject.service';

@Module({
  controllers: [SubjectController],
  providers: [SubjectService, SubjectRepository, PrismaService],
})
export class SubjectModule {}
