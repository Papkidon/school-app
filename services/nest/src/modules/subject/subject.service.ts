import { Injectable } from '@nestjs/common';
import { SubjectRepository } from './subject.repository';
import { subjectType } from '../../models/subject/subject.types';
import { DeepPartial } from '../../models/utility.types';
import { Subject } from '@prisma/client';
import { subjectsWithClassesDto } from '../../models/subject/subject.types';

@Injectable()
export class SubjectService {
  constructor(private subjectRepository: SubjectRepository) {}
  async findAll(): Promise<Subject[]> {
    return await this.subjectRepository.findAll();
  }

  async findById(id: string): Promise<Subject> {
    return await this.subjectRepository.findById(id);
  }

  async findByName(name: string): Promise<Subject> {
    return await this.subjectRepository.findByName(name);
  }

  async findAllSubjectWithClasses(): Promise<subjectsWithClassesDto> {
    return await this.subjectRepository.findAllSubjectsWithClasses();
  }

  async create(body: any): Promise<Subject> {
    return await this.subjectRepository.create(body);
  }

  async delete(id: string): Promise<Subject> {
    return await this.subjectRepository.delete(id);
  }

  async update(id: string, body: DeepPartial<subjectType>): Promise<Subject> {
    return await this.subjectRepository.update(id, body);
  }
}
