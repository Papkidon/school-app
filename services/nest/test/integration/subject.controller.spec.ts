import { INestApplication } from '@nestjs/common';
import { Test } from '@nestjs/testing';
import * as request from 'supertest';
import { subjectsWithClassesDto } from '../../src/models/subject/subject.types';
import { SubjectModule } from '../../src/modules/subject/subject.module';
import { SubjectRepository } from '../../src/modules/subject/subject.repository';
import {
  dummySubjectCreate,
  dummySubjectsWithClasses,
} from './dummies/dummy.data';

describe('[SubjectController] Integration tests', () => {
  let app: INestApplication;
  const subjectRepository = {
    findAll: () => ['dummyFoundSubjects'],
    findById: () => 'dummyFoundSubjectById',
    findByName: () => 'dummyFoundSubjectByName',
    findAllSubjectsWithClasses: () =>
      dummySubjectsWithClasses as unknown as subjectsWithClassesDto,
    create: () => 'dummyCreatedSubject',
    update: () => 'dummyUpdatedSubject',
    delete: () => 'dummyDeletedSubject',
  };

  beforeAll(async () => {
    const moduleRef = await Test.createTestingModule({
      imports: [SubjectModule],
    })
      .overrideProvider(SubjectRepository)
      .useValue(subjectRepository)
      .compile();

    app = moduleRef.createNestApplication();
    await app.init();
  });

  it(`Get all subjects - should pass when at least one subject exists in the db`, async () => {
    const response = await request(app.getHttpServer()).get('/api/v1/subject/');

    expect(response.status).toEqual(200);
    expect(response.body).toMatchSnapshot();
  });

  it(`Get all subjects - should fail when no subject data exists in the db`, async () => {
    jest.spyOn(subjectRepository, 'findAll').mockImplementationOnce(() => []);

    const response = await request(app.getHttpServer()).get('/api/v1/subject/');

    expect(response.status).toEqual(404);
    expect(response.body).toMatchSnapshot();
  });

  it(`Get subject by id - should pass when subject with given id exists in the db`, async () => {
    const response = await request(app.getHttpServer()).get(
      '/api/v1/subject/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    );

    expect(response.status).toEqual(200);
    expect(response.body).toMatchSnapshot();
  });

  it(`Get subject by id - should fail when subject with given id does not exists in the db`, async () => {
    jest.spyOn(subjectRepository, 'findById').mockImplementationOnce(() => '');

    const response = await request(app.getHttpServer()).get(
      '/api/v1/subject/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    );

    expect(response.status).toEqual(404);
    expect(response.body).toMatchSnapshot();
  });

  it('Get subject by id - should fail when id parameter is not UUID', async () => {
    const response = await request(app.getHttpServer()).get(
      '/api/v1/subject/123',
    );

    expect(response.status).toEqual(422);
    expect(response.body).toMatchSnapshot();
  });

  it('Get subjects with classes - should pass when data exists', async () => {
    const response = await request(app.getHttpServer()).get(
      '/api/v1/subject/classes',
    );

    expect(response.status).toEqual(200);
    expect(response.body).toMatchSnapshot();
  });

  it('Get subjects with classes - should fail when data does not exists', async () => {
    jest
      .spyOn(subjectRepository, 'findAllSubjectsWithClasses')
      .mockImplementationOnce(() => '' as unknown as subjectsWithClassesDto);

    const response = await request(app.getHttpServer()).get(
      '/api/v1/subject/classes',
    );

    expect(response.status).toEqual(404);
    expect(response.body).toMatchSnapshot();
  });

  it(`Create subject - should pass when request body is correct`, async () => {
    jest
      .spyOn(subjectRepository, 'findByName')
      .mockImplementationOnce(() => '');

    const response = await request(app.getHttpServer())
      .post('/api/v1/subject/')
      .send(dummySubjectCreate);

    expect(response.status).toEqual(201);
    expect(response.body).toMatchSnapshot();
  });

  it(`Create subject - should fail when subject with given number already exists`, async () => {
    const response = await request(app.getHttpServer())
      .post('/api/v1/subject/')
      .send(dummySubjectCreate);

    expect(response.status).toEqual(409);
    expect(response.body).toMatchSnapshot();
  });

  it('Create subject - should fail when request body is incorrect', async () => {
    const response = await request(app.getHttpServer())
      .post('/api/v1/subject/')
      .send({ wrong: 'body' });

    expect(response.status).toEqual(422);
    expect(response.body).toMatchSnapshot();
  });

  it(`Update subject - should pass when subject with given id exists in the db`, async () => {
    jest
      .spyOn(subjectRepository, 'findById')
      .mockImplementationOnce(() => 'dummyFoundSubjectToUpdate');

    const response = await request(app.getHttpServer()).patch(
      '/api/v1/subject/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    ); // It passes without the body, because every property is optional

    expect(response.status).toEqual(200);
    expect(response.body).toMatchSnapshot();
  });

  it(`Update subject - should fail when subject with given id does not exists in the db`, async () => {
    jest.spyOn(subjectRepository, 'findById').mockImplementationOnce(() => '');

    const response = await request(app.getHttpServer()).patch(
      '/api/v1/subject/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    ); // It passes without the body, because every property is optional

    expect(response.status).toEqual(404);
    expect(response.body).toMatchSnapshot();
  });

  it(`Update subject - should fail when id parameter is not UUID`, async () => {
    const response = await request(app.getHttpServer()).patch(
      '/api/v1/subject/123',
    ); // It passes without the body, because every property is optional

    expect(response.status).toEqual(422);
    expect(response.body).toMatchSnapshot();
  });

  it(`Delete subject - should pass when subject with given id exists in the db`, async () => {
    const response = await request(app.getHttpServer()).delete(
      '/api/v1/subject/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    );

    expect(response.status).toEqual(200);
    expect(response.body).toMatchSnapshot();
  });

  it(`Delete subject - should fail when subject with given id does not exist in the db`, async () => {
    jest.spyOn(subjectRepository, 'findById').mockImplementationOnce(() => '');

    const response = await request(app.getHttpServer()).delete(
      '/api/v1/subject/e4067cdc-dc91-41a6-bbb3-650fdd352dd4',
    );

    expect(response.status).toEqual(404);
    expect(response.body).toMatchSnapshot();
  });

  it(`Delete subject - should fail when id parameter is not UUID`, async () => {
    const response = await request(app.getHttpServer()).delete(
      '/api/v1/subject/123',
    );

    expect(response.status).toEqual(422);
    expect(response.body).toMatchSnapshot();
  });

  afterAll(async () => {
    await app.close();
  });
});
