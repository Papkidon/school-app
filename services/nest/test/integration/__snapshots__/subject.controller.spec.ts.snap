// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[SubjectController] Integration tests Create subject - should fail when request body is incorrect 1`] = `
[
  {
    "code": "invalid_type",
    "expected": "object",
    "message": "Required",
    "path": [
      "data",
    ],
    "received": "undefined",
  },
  {
    "code": "unrecognized_keys",
    "keys": [
      "wrong",
    ],
    "message": "Unrecognized key(s) in object: 'wrong'",
    "path": [],
  },
]
`;

exports[`[SubjectController] Integration tests Create subject - should fail when subject with given number already exists 1`] = `
{
  "message": "Subject with given name already exists",
  "statusCode": 409,
}
`;

exports[`[SubjectController] Integration tests Create subject - should pass when request body is correct 1`] = `{}`;

exports[`[SubjectController] Integration tests Delete subject - should fail when id parameter is not UUID 1`] = `
[
  {
    "code": "invalid_string",
    "message": "Invalid uuid",
    "path": [],
    "validation": "uuid",
  },
]
`;

exports[`[SubjectController] Integration tests Delete subject - should fail when subject with given id does not exist in the db 1`] = `
{
  "message": "Subject with given id does not exist in the database",
  "statusCode": 404,
}
`;

exports[`[SubjectController] Integration tests Delete subject - should pass when subject with given id exists in the db 1`] = `{}`;

exports[`[SubjectController] Integration tests Get all subjects - should fail when no subject data exists in the db 1`] = `
{
  "message": "No subject data exists in the database",
  "statusCode": 404,
}
`;

exports[`[SubjectController] Integration tests Get all subjects - should pass when at least one subject exists in the db 1`] = `
[
  "dummyFoundSubjects",
]
`;

exports[`[SubjectController] Integration tests Get subject by id - should fail when id parameter is not UUID 1`] = `
[
  {
    "code": "invalid_string",
    "message": "Invalid uuid",
    "path": [],
    "validation": "uuid",
  },
]
`;

exports[`[SubjectController] Integration tests Get subject by id - should fail when subject with given id does not exists in the db 1`] = `
{
  "message": "Subject with given id does not exist in the database",
  "statusCode": 404,
}
`;

exports[`[SubjectController] Integration tests Get subject by id - should pass when subject with given id exists in the db 1`] = `{}`;

exports[`[SubjectController] Integration tests Get subjects with classes - should fail when data does not exists 1`] = `
{
  "message": "No subject with classes data was found in the database",
  "statusCode": 404,
}
`;

exports[`[SubjectController] Integration tests Get subjects with classes - should pass when data exists 1`] = `
[
  {
    "classes": [
      {
        "created_at": "2023-03-11T14:19:21.154",
        "id": "d379d73a-7958-44e4-ab59-0e4128f35295",
        "max_population": 20,
        "number": 211,
        "updated_at": "2023-03-11T14:19:21.154",
      },
    ],
    "id": "917d669d-ea51-4031-b7ef-25b1850bbd52",
    "name": "geography",
    "points": 1,
  },
  {
    "classes": [
      {
        "created_at": "2023-03-11T14:19:21.166",
        "id": "261133c7-c73c-4a7f-8199-548798dc0e97",
        "max_population": 15,
        "number": 111,
        "updated_at": "2023-03-11T14:19:21.166",
      },
    ],
    "id": "c3114bc1-1bfa-4f44-ba17-f199a6dc635c",
    "name": "math",
    "points": 5,
  },
]
`;

exports[`[SubjectController] Integration tests Update subject - should fail when id parameter is not UUID 1`] = `
[
  {
    "code": "invalid_string",
    "message": "Invalid uuid",
    "path": [],
    "validation": "uuid",
  },
]
`;

exports[`[SubjectController] Integration tests Update subject - should fail when subject with given id does not exists in the db 1`] = `
{
  "message": "Subject with given id does not exist in the database",
  "statusCode": 404,
}
`;

exports[`[SubjectController] Integration tests Update subject - should pass when subject with given id exists in the db 1`] = `{}`;
